### ISSUE TITLE
Carlini methods do not affect the images [Keras/Tensorflow]

### ISSUE THREAD

**Issue by marcoancona** [IBM/adversarial-robustness-toolbox#101]
**Describe the bug**
It seems that Carlini methods do not produce any adversarial examples using Keras 2.2.4 and Tensorflow backend 1.12.0 (adversarial example are identical to original examples)

**To Reproduce**
Use Keras 2.2.4 and Tensorflow 1.12.
Run the example reported below.

**Expected behavior**
Accuracy of the model evaluated with adversarial example should be lower.

**Screenshots**
![image](https://user-images.githubusercontent.com/2278117/60602219-aeddf780-9db3-11e9-9ff3-09eadfc146da.png)

**System information (please complete the following information):**
 - Ubuntu 16.04
 - Python 3.6
 - ART version or commit number: 267a6e6  
 - TensorFlow 1.12 / Keras 2.2.4

Test code
```
'''Trains a simple convnet on the MNIST dataset.

Gets to 99.25% test accuracy after 12 epochs
(there is still a lot of margin for parameter tuning).
16 seconds per epoch on a GRID K520 GPU.
'''

from __future__ import print_function
import keras
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv2D, MaxPooling2D
from keras import backend as K
import numpy as np

batch_size = 128
num_classes = 10
epochs = 1

# input image dimensions
img_rows, img_cols = 28, 28

# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()

if K.image_data_format() == 'channels_first':
    x_train = x_train.reshape(x_train.shape[0], 1, img_rows, img_cols)
    x_test = x_test.reshape(x_test.shape[0], 1, img_rows, img_cols)
    input_shape = (1, img_rows, img_cols)
else:
    x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
    x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
    input_shape = (img_rows, img_cols, 1)

x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
print('x_train shape:', x_train.shape)
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

# convert class vectors to binary class matrices
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)

model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])

model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

# Start attack section
from art.classifiers import KerasClassifier
from art.attacks import CarliniL2Method, CarliniLInfMethod

classifier = KerasClassifier(model, use_logits=False)
attacker = CarliniL2Method(classifier, targeted=False, confidence=100, max_iter=5, learning_rate=0.1)
x_test_adv = attacker.generate(x_test[:100])
score = model.evaluate(x_test_adv, y_test[:100], verbose=0)
print("L2 dist adv: ", np.sum((x_test_adv-x_test[:100])**2))
print('Adv. Test accuracy:', score[1])
```

**Comment by beat-buesser** [comment]
Hi @marcoancona Thank you for opening this issue and for the detailed test script. We will try to reproduce and fix it as soon as possible.

**Comment by mathsinn** [comment]
@marcoancona One thing that strikes me in your example is the very large `confidence` value and relatively small number for `max_iter`. Have you tried different values? @beat-buesser I suggest to also try that in your inquiry.

**Comment by marcoancona** [comment]
Yes I tried different combinations and up to 40 iterations. I also tried the default parameters (but with a non-targeted attack) and the result seems to be the same

**Comment by beat-buesser** [comment]
@marcoancona @mathsinn I did some runs of the test script above. My only changes are reduceing the `confidence` for different `max_iter` and I got the results below. It looks like `CarliniL2Method` works fine.

`attacker = CarliniL2Method(classifier, targeted=False, confidence=0, max_iter=50, learning_rate=0.1)`
L2 dist adv:  902.9391
Adv. Test accuracy: 0.11

`attacker = CarliniL2Method(classifier, targeted=False, confidence=0, max_iter=5, learning_rate=0.1)`
L2 dist adv:  852.49304
Adv. Test accuracy: 0.25

`attacker = CarliniL2Method(classifier, targeted=False, confidence=1, max_iter=5, learning_rate=0.1)`
L2 dist adv:  1088.6107
Adv. Test accuracy: 0.27

`attacker = CarliniL2Method(classifier, targeted=False, confidence=10, max_iter=50, learning_rate=0.1)`
L2 dist adv:  717.30115
Adv. Test accuracy: 0.75

**Comment by marcoancona** [comment]
Oh, indeed it works for me as well with these parameters. Sorry, my bad for not having done a more exhaustive parameters search. On the other hand, the [documentation](https://adversarial-robustness-toolbox.readthedocs.io/en/latest/modules/attacks.html#carlini-and-wagner-l-2-attack) says 

> confidence (float) – Confidence of adversarial examples: **a higher value produces examples that are farther away, from the original input, but classified with higher confidence as the target class**.

I see the exact opposite behavior running this test:
```py
for confidence in [0, 0.5, 1, 2, 4, 8, 16, 32, 64]:
    attacker = CarliniL2Method(classifier, targeted=False, confidence=confidence, max_iter=10, learning_rate=0.1)
    x_test_adv = attacker.generate(x_test[:100])
    score = model.evaluate(x_test_adv, y_test[:100], verbose=0)
    print("Confidence: ", confidence)
    print("L2 dist adv: ", np.sum((x_test_adv-x_test[:100])**2))
    print('Adv. test accuracy:', score[1])
    print('')
```
```
Confidence:  0
L2 dist adv:  1009.6896
Adv. Test accuracy: 0.08

Confidence:  0.5
L2 dist adv:  1102.9342
Adv. Test accuracy: 0.09

Confidence:  1
L2 dist adv:  1197.8207
Adv. Test accuracy: 0.1

Confidence:  2
L2 dist adv:  1306.3167
Adv. Test accuracy: 0.15

Confidence:  4
L2 dist adv:  1399.4993
Adv. Test accuracy: 0.31

Confidence:  8
L2 dist adv:  907.717
Adv. Test accuracy: 0.68

Confidence:  16
L2 dist adv:  0.0
Adv. Test accuracy: 1.0

Confidence:  32
L2 dist adv:  0.0
Adv. Test accuracy: 1.0

Confidence:  64
L2 dist adv:  0.0
Adv. Test accuracy: 1.0
```
Does the effect of `confidence` depend on the other parameters as well? In this case, I would point that out in the documentation.
